proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/InMemJournal.scala

```diff
@@ -0,0 +1,220 @@
+
+object InmemJournal {
+  case class EventsByTag(tag: String, fromOffset: Long)
+}
+
+private[persistence] class InmemJournal extends AsyncWriteJournal with InmemMessages with ActorLogging {
+
+  import InmemJournal.EventsByTag
+
+  override implicit protected lazy val mat: Materializer = ActorMaterializer()
+
+  override def asyncWriteMessages(messages: immutable.Seq[AtomicWrite]): Future[immutable.Seq[Try[Unit]]] = {
+    for (w <- messages; p <- w.payload)
+      add(p)
+    Future.successful(Nil) // all good
+  }
+
+  override def asyncReadHighestSequenceNr(persistenceId: String, fromSequenceNr: Long): Future[Long] =
+    Future.successful(highestSequenceNr(persistenceId))
+
+  override def asyncReplayMessages(persistenceId: String, fromSequenceNr: Long, toSequenceNr: Long, max: Long)(
+      recoveryCallback: PersistentRepr => Unit
+  ): Future[Unit] = {
+    val highest = highestSequenceNr(persistenceId)
+    if (highest != 0L && max != 0L)
+      read(persistenceId, fromSequenceNr, math.min(toSequenceNr, highest), max).foreach(recoveryCallback)
+    Future.successful(())
+  }
+
+  override def asyncDeleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
+    val toSeqNr = math.min(toSequenceNr, highestSequenceNr(persistenceId))
+    var snr = 1L
+    while (snr <= toSeqNr) {
+      delete(persistenceId, snr)
+      snr += 1
+    }
+    Future.successful(())
+  }
+
+  override def receivePluginInternal: Receive = {
+    case EventsByTag(tag, fromOffset) =>
+      log.info("Received EventsByTag query for tag {} from offset {}", tag, fromOffset)
+      sender() ! eventsByTagQuery(tag, fromOffset)
+  }
+}
+
+class InmemReadJournal(system: ExtendedActorSystem, config: Config)
+    extends EventsByTagQuery
+    with ReadJournalProvider
+    with ReadJournal
+    with akka.persistence.query.javadsl.ReadJournal {
+
+  private def inmemJournal: ActorRef = Persistence(system).journalFor("inmem-journal")
+  private implicit val timeout = Timeout(5.seconds)
+  import akka.pattern.ask
+  import InmemJournal._
+
+  override def scaladslReadJournal(): ReadJournal = this
+
+  override def javadslReadJournal(): akka.persistence.query.javadsl.ReadJournal =
+    // todo this is obviously not right
+    this
+
+  override def eventsByTag(tag: String, offset: Offset): Source[EventEnvelope, NotUsed] = {
+    val fromOffset = offset match {
+      case Sequence(seq) => seq
+      case NoOffset => 0L
+      case unsupported =>
+        throw new IllegalArgumentException(s"$unsupported is an unsupported offset type for the in memory read journal")
+    }
+
+    Source
+      .fromFutureSource(
+        (inmemJournal ? EventsByTag(tag, fromOffset)).mapTo[Source[EventEnvelope, NotUsed]]
+      )
+      .mapMaterializedValue(_ => NotUsed)
+  }
+}
+
+/**
+ * INTERNAL API.
+ */
+trait InmemMessages {
+  private var messages = Map.empty[String, Vector[Message]]
+  private var allMessages = Vector.empty[Message]
+  private var offset: Long = 0
```

Comment:
Do we need these vars to be thread-safe?

Reply:
Yeah, I'll mark them as volatile.

Reply:
Actually no, they are only accessed from within an actor.

Reply:
Ok, cool.